design: cloud change logs

jbw976 opened the pull request:
### Description of your changes

This PR contains a 1-pager design document for the cloud change logs feature described in #5477.  The alpha implementation of this feature is intended to be included in the v1.17 release.

This design stems from some collaborative brainstorming sessions with @negz, so thank you for the assistance so far! 🙇‍♂️ 

While this design is being reviewed by the maintainer team and community, I will be prototyping on the side to a) further validate the feasibility of the proposed design and b) start making progress without delay.

Pro-tip to reviewers: this document makes heavy usage of markdown section headers, as well as frequent summary checkpoints for the reader, so the document is easily parseable and navigable using github's "outline" view:
![Screenshot 2024-07-08 at 10 09 25 PM](https://github.com/crossplane/crossplane/assets/4313439/f7f4a67b-9523-4ec0-9aa1-47cf3b51586a)

I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+loss of these logs. Change logs entries will be written to a local location
+immediately, so that a healthy network connection is not required at the time an
+entry is generated. This local location doesn't necessarily need to be a highly
+durable and persistent location itself, they key is that we don't lose any data
+if we temporarily lose network connectivity while entries are being generated.
```
```suggestion
durable and persistent location itself, the key is that we don't lose any data
if we temporarily lose network connectivity while entries are being generated.
```

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+pod logs, and they can also be exported to external/centralized systems for long
+term storage and further inspection and analysis. One suggested means to collect
+and export the change logs is to use the [OpenTelemetry
+Collector](https://opentelemetry.io/docs/collector/), but a more details
+analysis of that setup is out of scope for this document.
```
```suggestion
Collector](https://opentelemetry.io/docs/collector/), but a more detailed
analysis of that setup is out of scope for this document.
```

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+
+As mentioned in the previous section, change log entries will be written to the
+pod logs. However, there is some additional structure we will build in order to
+keep change logs entirely separated from the regular provider log entries, which
+will add not only with human readability, but also post processing and analysis
```
```suggestion
keep change logs entirely separated from the regular provider log entries, which
will aid not only with human readability, but also post processing and analysis
```

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+gRPC connection between the two containers will be established over a Unix
+domain socket on a shared `emptyDir` volume mounted by both containers. This
+setup allows the main provider container to send the change log entries to the
+sidecar container in a fairly performant manner without needed any network
+access at all.
```
```suggestion
sidecar container in a fairly performant manner without needing any network
access at all.
```

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+* The provider will have sidecar container where each entry is sent
+* Main container will communicate through gRPC with the sidecar container over a
+  Unix domain socket on a shared volume
+* Data is serialized in a protobuf binary format and written to sidecar
+  container logs in JSON format
```
```suggestion
* Data is serialized in a protobuf binary format and written to the sidecar
  container logs in JSON format
```

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+[alpha](https://docs.crossplane.io/latest/learn/feature-lifecycle/#alpha-features)
+level of maturity when it is first released, meaning that Crossplane users must
+opt-in to using this functionality by setting an `--enable-change-logs` feature
+flag. Note this will flag will need to be set on each provider for which change
+logs are desired, likely via a `DeploymentRuntimeConfig`.
```
```suggestion
flag. Note this flag will need to be set on each provider for which change
logs are desired, likely via a `DeploymentRuntimeConfig`.
```

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+flag. Note this will flag will need to be set on each provider for which change
+logs are desired, likely via a `DeploymentRuntimeConfig`.
+
+Also note in the plan below that core Crossplane itself does not need to updated
+at all to support this feature. The key changes are in `crossplane-runtime`, the
+Upjet framework, and the ecosystem of providers.
```
```suggestion
Also note in the plan below that core Crossplane itself does not need to be updated
at all to support this feature. The key changes are in `crossplane-runtime`, the
Upjet framework, and the ecosystem of providers.
```

jeanduplessis commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+* The [Upjet framework](https://github.com/crossplane/upjet) is updated to use
+  this new `crossplane-runtime` logic and with additional code generation logic
+  to generate providers that:
+  * Create a initialize a sidecar container in the provider pod
+  * Start and initialize the gRPC server in the sidecar container and the gRPC
+    client in the main provider container
```
```suggestion
  * Create and initialize a sidecar container in the provider pod
  * Start and initialize the gRPC server in the sidecar container and the gRPC
    client in the main provider container
```

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+Crossplane is deployed to manage the most critical infrastructure resources for
+many of the organizations that have chosen to adopt it. Crossplane acts in a
+continuously reconciling loop to drive the state of external systems and
+resources to match the desired state declared by the users platform. If changes
```
```suggestion
resources to match the desired state declared by the users' platform. If changes
```

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+systems even when no user is actively interacting with the control plane.
+Consider the scenario when someone updates a resource outside of Crossplane,
+e.g. via the AWS console or `gcloud` CLI. When Crossplane detects this
+"configuration drift", it will enforce its "source of truth" to immediately
```
```suggestion
"configuration drift", it will enforce its "source of truth" to eventually
```
;-)

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+With Crossplane acting continuously and autonomously to update critical
+infrastructure, it is vital for users to have insight into the operations being
+performed, so they can build and maintain a strong sense of confidence and trust
+in their control planes.
```
This thought comes to mind, playing devil's advocate: if this feature is groundbreaking and critical, why not just look at cloud provider audit logs? They exist. Why are they not enough?

negz replied:
> Why are they not enough?

Compared to cloud change logs:

* Not all systems Crossplane has providers for will have audit logs.
* Cloud change logs are homogenous, so it's easier to configure them all to go to some single pane of glass (compared to configuring Google audit logs, AWS audit logs, etc).
* Cloud change logs take away any guesswork as to what changes Crossplane made, as opposed to trying to use cloud identity, or HTTP user agents etc to identify Crossplane changes.

You can probably get pretty far just setting up cloud audit logs, but I don't think that negates the value of this feature.

sttts replied:
All good arguments. Why not write it down? :)

jbw976 replied:
Added a new alternatives considered section for this cloud providers audit logs idea too 🤓

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+|-------------------| ------- |
+| Timestamp | ISO 8601 format, e.g. `2023-04-01T12:34:56Z` |
+| Provider name | `xpkg.upbound.io/upbound/provider-aws-ec2:v1.8.0` |
+| Managed Resource identifying data (GVK/name) | `{apiVersion: ec2.aws.upbound.io/v1beta2, kind: Instance, name: dev-instance-bt66d}` |
```
RV ?

sttts replied:
Nvmd. Will be part of the JSON blob below. Admittedly, also the upper "Managed Resource identifying data" would be redundant.

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+compared to observing the external system and recording the "after" state
+immediately after the operation as well. If we find during implementation and
+testing that recording the "after" state places too much burden on the system,
+we can revisit this decision.
```
Some data point: we once did a benchmark on OpenShift with API auditing on with essentially the same data as you describe and measured 10% degradation.

sttts replied:
yes, upstream kube audit logs with full body of requests and responses.

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+The Kubernetes [audit
+logs](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/) capture a
+similar concept to the change logs feature we are proposing, but are not a great
+fit for our goals, both in the data that it captures and the scenarios for which
```
```suggestion
fit for our goals, both in the data that it captures and the scenarios for which
```

I don't understand this statement. The data is the same, with the exception of the optional extra JSON blob.

sttts replied:
Capturing data directly related to provider actions with more control about the when and why certainly has value. So there is an advantage of the controller-runtime based approach. Am just saying here that the actual data captured (modulo the additional provider-specific blob) is the same.

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+of a configuration drift occurs during a regular polling `Reconcile()` call,
+which is not in response to any request at the API server level. This scenario
+would not be captured in the Kubernetes audit logs, as no API call was made to
+the Kubernetes API server. Therefore, Kubernetes audit logs are not a great fit
```
Crossplane would persist the temporary `status.atProvider`, which means that there is an API audit event.

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+
+The managed reconciler calls the provider's [`ExternalClient.Observe()`
+method](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L914)
+before any CRUD operation is performed, in order to understand the current state
```
CUD to be precise :)

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+* [Update](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L1189)
+* [Delete](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L954)
+
+The code path through the managed reconciler's `Reconcile()` method are already
```
```suggestion
The code paths through the managed reconciler's `Reconcile()` method are already
```

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+a bit complex. One design choice that has helped minimize the complexity and
+assist in readability is that the reconciler returns as early as it can after
+performing an operation. Since we will now be adding an additional `Observe()`
+call after each CRUD operation, this will increase the complexity further, but
```
don't cloud providers return the new state of the world after a CRUD operation? How many do?

sttts commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+We consider change logs entries to be related to both security and reliability
+of the control plane, and therefore we will have a fairly low tolerance for data
+loss of these logs. Change logs entries will be written to a local location
+immediately, so that a healthy network connection is not required at the time an
```
```suggestion
immediately, so that a healthy network connection (beyond the pod boundaries, grpc must work) is not required at the time an
```

sttts replied:
nvmd. Unix domain sockets as you describe below :+1:

ulucinar commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+| Provider name | `xpkg.upbound.io/upbound/provider-aws-ec2:v1.8.0` |
+| Managed Resource identifying data (GVK/name) | `{apiVersion: ec2.aws.upbound.io/v1beta2, kind: Instance, name: dev-instance-bt66d}` |
+| Operation Type | `create\|update\|delete` |
+| Before operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider` |
+| After operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider`  |
```
For providers returning the control back to the managed reconciler before the external operation completes (upjet does so in the default async reconciliation mode), we will in general record the same state as the `before` and the `after` state. We were planning to introduce modifications in the managed reconciler so that async reconciliation becomes a first class citizen in crossplane-runtime. Another option could be to give the control of logging to the provider itself, e.g., give a handle of the gRPC client to the provider, so that the upjet runtime can properly report the before-update & after-update states of the external resource.

We also discussed the utility of the `spec.forProvider` as the reconciler does not get a hold of the old spec in case it's been updated (this would especially be relevant if it's the spec update that will result in an external resource update and in turn, result in a change log). The only common exception is late-initialization but this is not relevant to the change logs discussion. So, the desired state will in general be identical in the before and after logs related to a single change.

negz replied:
> We were planning to introduce modifications in the managed reconciler so that async reconciliation becomes a first class citizen in crossplane-runtime.

Is there somewhere I can learn more about these plans? I'd like to understand the async operation approach better. I understand that Terraform operations can take a long time to complete, but I've never personally been sold on the async operation approach.

jbw976 replied:
Thanks @ulucinar for writing down this observation about providers that perform async operations, it's a good point to understand. As I've been thinking through this design further, I've been increasingly thinking of the "after" state we capture as being a "best effort" approach.

Within a single `Reconcile()`, we can perform the operation and then call `Observe()` to get the latest state from the provider, with the understanding that the available state at this point in time may not yet reflect the operation we just performed as being concluded. If there are no further change logs for the given resource, we can perhaps assume that it converged with the desired state.

That may be acceptable IMO, at least for this initial alpha implementation 🤔

turkenh replied:
> I understand that Terraform operations can take a long time to complete, but I've never personally been sold on the async operation approach.

Things might have changed with the latest changes in the upjet framework, especially with the no fork architecture, but AFAIR, a terraform apply call can take as long as the timeouts defined for a given terraform resource. For example, AWS [RDS instance](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/rds_cluster#timeouts) can take up to 120m and blocking the reconcile for long-running tasks is against controller [best practices](https://github.com/kubernetes-sigs/controller-runtime/issues/1763#issuecomment-1003703642). 

In provider-helm, we had a similar [need](https://github.com/crossplane-contrib/provider-helm/issues/63) and could live up to today without implementing it at the cost of not providing a great UX for some couple of rare cases where a helm install taking much longer (e.g. a pre-install hook, or wait=true). In case of terraform resources, this is more severe (i.e. longer).

jbw976 replied:
See https://github.com/crossplane/crossplane/pull/5822#discussion_r1701017542 for where we have landed on no longer collecting the "after" state by performing a 2nd `Observe()` call.

ulucinar commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+| Before operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider` |
+| After operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider`  |
+| Result of operation | `success` or error object |
+| (optional) Additional information that Providers can set as they choose | `{JSON object of arbitrary data}` |
```
I remember there was a separate discussion with @luxas on whether we can use the change logs feature to record data about all the reconciliations (for example, reconciliation timestamps), and not just about those reconciliations which attempt to correct drifts. Then the after observation would become optional (only recorded for those reconciliations that update the external resource). crossplane-runtime may not need to be opinionated on this, we would extend crossplane-runtime, and if needed, the package manager and the like, so that the side car (the gRPC server) and the client are properly setup, and the provider could use this infrastructure to implement some related features. The proposed arbitrary data field opens up the possibility for similar cases but the crossplane-runtime had better be un-opinionated about them, if we would like to allow other use cases.

Re-observing for the after-update external resource state would be the issue in this case, if we take the control from the managed reconciler. I think we would not like to have the external client to make a second observation on its own. One idea could be to allow the external client to be able to signal to the managed reconciler the need for an immediate reconciliation so that the managed reconciler requeues an immediate rate limited reconciliation request without waiting for the next poll period.

How do you folks feel about allowing the providers to record more reconciliation data, not limited to updates?

negz replied:
> How do you folks feel about allowing the providers to record more reconciliation data, not limited to updates?

I don't feel like I understand well enough to answer the question. What are some examples of reconciliation data other than timestamps?

jbw976 commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,310 @@
+
+Also note in the plan below that core Crossplane itself does not need to updated
+at all to support this feature. The key changes are in `crossplane-runtime`, the
+Upjet framework, and the ecosystem of providers.
```
Thank you very much for pointing this out @sttts! You opened my eyes to the fact that I made a naive assumption about how the upjet providers work and in general how a provider can influence its pod spec.  spoiler alert: it can't 😂 

I had assumed that at build time a provider specifies details about its pod spec, including its main container and any init or sidecar containers it wants as well. This is not true. The package manager is what crafts the provider's deployment/pod spec.  I had forgotten this 😊 

So I will definitely be looking into this in more detail, but some early thoughts:

* I'll see what's possible with a `DeploymentRuntimeConfig`. Perhaps a sidecar container snippet can be specified in there and then it gets merged nicely with other config the package manager will enforce 🙏 
* If not, the package manager could be updated to inject a sidecar container for each provider where change logs are desired. This would indeed require a code change to core crossplane, so thanks again for pointing this gap out @sttts!! 🙇‍♂️

jbw976 replied:
I was really pleased to see how well a `DeploymentRuntimeConfig` could be used to set this architecture up for a provider during my prototyping efforts last week. I've made the recommendation that at least for this alpha implementation, we can just use a `DeploymentRuntimeConfig` instead of having to make any changes to core crossplane and/or its package manager. I've also included an example of what this `DeploymentRuntimeConfig` looks like.

jbw976 reviewed (commented):
Thanks a lot for the detailed review @sttts! I have some more things to think through (e.g. how the sidecar pod could be injected and the format of the data transfer on the wire), but I've included responses to all of your comments! 🙇

jbw976 commented:
Thanks folks for the reviews and discussions so far, really appreciate it!

I spent some cycles last week prototyping an end to end effort to better understand the viability of the proposed design and how all the components fit together. I've responded to most all comments now, and I've pushed a few new commits with additional and clarifying content for:

* `DeploymentRuntimeConfig` works great to inject sidecar container,  we can rely on this instead of making changes to package manager. An example was added of what this looks like.
* Updated the scope and description of code changes we will make across the various Crossplane ecosystem components in our implementation rollout.
* Added further details about why Kubernetes audit logs are not an ideal fit for our needs.
* Added an alternative considered section for cloud provider audit logs too.

blakeromano commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+control plane when possible. For example, we do not want to build any logic to
+perform log rotation, compression, or other log management tasks.
+
+Change log entries will be written to `stdout` so they can be captured by the
```
I think we need to have a way for folks to configure their output. I think these logs in large environments will become unruly and super noisy where we may want to have folks say "write them to a seperate file on the node rather than stdout" because my fear here is I want to enable them but I may want them to go through a separate pathway for cost reasons and stored separately or for compliance reasons, retention on these logs may be different.

I get that can be done through a log collector like Otel, Fluentbit, etc; but I think that we will want to have a way to distinguish where the logs be sent out.

Maybe this could be done with a potential additional sidecar like fluentbit that just grabs the logs and then allow folks in a `deploymentruntimeconfig` to specify their fluentbit config for just those sets of logs?

negz replied:
>  I think these logs in large environments will become unruly and super noisy where we may want to have folks say "write them to a seperate file on the node rather than stdout" 

I'd explicitly like to avoid this. It starts to imply configuring log compression, log rotation, etc. I'd prefer to let another tool handle that.

> because my fear here is I want to enable them but I may want them to go through a separate pathway for cost reasons and stored separately or for compliance reasons, retention on these logs may be different.

This is why I prefer writing them to stdout from a sidecar container, not the "main" provider process. This way you can configure your log processor of choice (fluent bit etc) to handle them however you like. The way you distinguish them is "logs coming from the change-logs sidecar container".

turkenh replied:
https://github.com/crossplane/crossplane/pull/5822#discussion_r1700092670 could be related here.

turkenh commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+
+## Alternatives Considered
+
+The follow alternative approaches have been considered for this change logs design:
```
```suggestion
The following alternative approaches have been considered for this change logs design:
```

turkenh commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+
+* The [template repo](https://github.com/crossplane/provider-template) for
+  classic providers (not generated with Upjet) should be updated to include a
+  sidecar container, gRPC connection, etc.
```
> should be updated to include a sidecar container

Do we really need any change in the provider codebase to include a sidecar container? Looks like this is done via `DeploymentRuntimeConfig`.

jbw976 replied:
Good catch! yep with the current approach of using a `DeploymentRuntimeConfig` the providers themselves do not need to be updated to set up the sidecar container. I'll update this.

turkenh commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+* The major upjet based providers (e.g. `provider-upjet-aws`,
+  `provider-upjet-gcp`, `provider-upjet-azure`, etc.)
+  * Updated with changes similar to the upjet provider template (consumes new
+    Upjet version and creates gRPC client to pass into its controllers)
```
A PR against at least one native / non upjet provider could be helpful. This could be used by the community as an example to add support to their provider.

jbw976 replied:
Agreed! Actually for development simplicity, I have been iterating on https://github.com/crossplane-contrib/provider-kubernetes/compare/main...jbw976:provider-kubernetes:change-logs.  So that will definitely be a PR that I open 😉 

I'll udpate this design doc to make that known too.

turkenh commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+The code paths through the managed reconciler's `Reconcile()` method are already
+a bit complex. One design choice that has helped minimize the complexity and
+assist in readability is that the reconciler returns as early as it can after
+performing an operation. Since we will now be adding an additional `Observe()`
```
Additional `Observe` means one more extra call to the cloud provider API and will increase probability of API ratelimiting. At least for `Create`s, we always ` queue` in the managed reconcilers and have a consecutive `Observe`, which we can leverage somehow.

jbw976 replied:
I definitely hear the concerns about the extra `Observe()` call. The usefulness of it is even being called into question with https://github.com/crossplane/crossplane/pull/5822#discussion_r1687250806 😊 

I'm not against cutting the "after" state completely, at least from this initial alpha implementation. I think we still tell an interesting story about WHY we made a change without having that data.

Not implementing an "after" state may be preferable instead of having to manage state across multiple reconciles to link together the `Observe()` results from a subsequent reconcile with the `Create()` or `Update()` operation from a previous reconcile. As things are currently implemented (https://github.com/crossplane/crossplane-runtime/compare/master...jbw976:crossplane-runtime:change-logs), everything is contained within a single reconcile and is definitely more simple for that 😇 

So in this alpha implementation, we could:

1. perform the extra `Observe()` call and see the feedback from early adopters
2. cut the "after" state functionality and do nothing
3. code something more complex that connects `Observe()` state to `Create()/Update()/Delete()` observations across reconciliations

Strong opinions from you all?

bobh66 replied:
I think alpha is a great time to see what the impact of the second `Observe()` call would be.  If it's bad we can always cut it when it goes beta.

bobh66 replied:
You could always put it under it's own nested flag so the feature could run with or without the second `Observe()` call.

jbw976 replied:
Chatted with @negz a bit more about "after" state and a 2nd `Observe()` call:

* From the first `Observe()` call, we should have a very up to date `status.atProvider` for any modern managed resource
* The difference between `spec.forProvider` and `status.atProvider` should fairly conclusively indicate why we are performing an operation
* The content of an "after" state collection from a 2nd `Observe()` is likely to not be any different from the "before", and therefore have limited value as called out in https://github.com/crossplane/crossplane/pull/5822#discussion_r1687250806

Therefore, it's reasonable to remove from this design the "after" state and thus any reason to perform a 2nd `Observe()`.  Without it, we still have a useful record of why we are making changes, whether that be because the spec's desired state has changed, we detected drift in the external resource, etc.

Perhaps Providers could be extended further in the future to provide more details in their [`ExternalObservation`](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L399) about why they concluded `ResourceUpToDate: false` and we can include that information in the change logs. For example, there is a `ExternalObservation.Diff` field, but I don't believe many Providers are currently setting it.

bobh66 commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+opt-in to using this functionality by setting an `--enable-change-logs` feature
+flag. Note this flag will need to be set on each provider for which change
+logs are desired, likely via a `DeploymentRuntimeConfig`.
+
```
Has there been any discussion of whether this feature will be enabled-by-default after it leaves alpha?  I can see where this will be very useful to a lot of people, but it also seems to (potentially) add a non-trivial amount of overhead that is likely to be unneeded for many use cases.  As long as the extra `Observe()` call is required I think we need to be able to turn off the feature, or require it to be specifically enabled even after alpha.

jbw976 replied:
Yes, this will definitely be an alpha feature that folks have to opt-in to enable it!  See https://github.com/jbw976/crossplane/blob/design-change-logs/design/one-pager-change-logs.md#setup-and-configuration

bobh66 reviewed (commented):
Overall I think it's looking good - my main concern is the extra Observe and the impact it has on the provider API usage/throttling.

jbw976 commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+  * Managed reconciler is updated to generate change log entries, send them to
+  the gRPC server (i.e. sidecar container), and to include additional calls to
+  `Observe()` after each CUD operation to capture the "after" state of the entry
+* `change-logs-sidecar`
+  * A new repository is created to build and publish the image used in the sidecar container
+  * Consumes the protobuf types from `crossplane-runtime`
+  * Starts the gRPC server and listens for change log entries
+  * Writes all entries to `stdout` (i.e. pod logs)
```
Interesting idea @turkenh - in general I like the idea of using standard tools to not have to build too much ourselves and to integrate with other systems more easily.

Do you know of any examples that show how to actually use the logs SDK? Just reading through the [spec](https://opentelemetry.io/docs/specs/otel/logs/sdk/) you linked, I'm already confused about how to actually make it work 😂. I'm not quickly finding practical examples of how to set up and configure these components, perhaps you already know where to find those? 

I've basically built the entire scenario described in this design document end to end already, see:
* https://github.com/crossplane/crossplane-runtime/compare/master...jbw976:crossplane-runtime:change-logs
* https://github.com/crossplane-contrib/provider-kubernetes/compare/main...jbw976:provider-kubernetes:change-logs
* https://github.com/jbw976/change-log-sidecar
  * a lot of these files are just for my dev/testing iterations 😇 

The code isn't too complex (but certainly needs polishing too), so the maintenance overhead over time isn't too daunting. I also like how the experience is very specific and scoped to the information we're trying to collect from crossplane-runtime and the providers, as opposed to getting something more heavy weight involved.

Your last 2 points are probably the most interesting (about being able to easily send to a central instance, and to integrate with processors easily). I had assumed that if someone chooses to set up the otel collector in their control plane, and have it easily gather the change log sidecar logs which are nicely isolated for them, they could use processors too. Do you think that isn't possible unless the log entries were originally emitted with the otel logs SDK?

How strongly do you feel about this idea? 🤔

negz replied:
I looked into the Otel logging approach before I suggested the design that @jbw976 has been running with. While I'd typically agree it's best not to reinvent the wheel,  I wasn't convinced we should use Otel logs (or really events) for this.

Quoting myself from Upbound's internal Slack a few weeks ago:

> I looked into otel (log) events a bit more this afternoon, as a possible way to emit change logs:
> 
> * https://opentelemetry.io/docs/specs/otel/logs/event-api/#emit-event
> * https://github.com/open-telemetry/opentelemetry-specification/issues/4045
> 
> They’re a bit oddball. At the RPC layer they’re just a log with no severity, an event.name attribute, and (by  convention) a structured body , with the same schema for all events with the same name. The Go SDK only has beta support for logging and has no explicit support for events (SDKs are supposed to have specific abstractions for events even though they’re just specially formatted logs). You could just manually emit logs that are by convention an event, but at the end of the day the SDKs just wrap some underlying logger (e.g. zap).

AFAICT:

* The Go logs SDK seems fairly immature. It's beta, and subject to breaking changes.
* We really want events (a special type of log), and there is no Go events SDK yet.
* The "structured" body of an Otel log isn't really structured - it's just an `Any` - https://github.com/open-telemetry/opentelemetry-proto/blob/v1.3.2/opentelemetry/proto/logs/v1/logs.proto#L170

I don't feel very strongly, but my intuition is that there's limited benefit to using Otel for this and that trying to fit into their model might make like more difficult than it needs to be.

negz replied:
> One can easily configure pushing logs to a central/shared otel collector instead, avoiding 1 sidecar per provider.

Whether we use Otel or not, I'd prefer to avoid _pushing_ logs to a centralized collector.

Pushing to a sidecar container was an intentional design choice. It gives us a local(ish) cache of change logs that a centralized collector can scrape. If we were to push logs to a centralized collector we'd risk losing change logs if a provider was partitioned from the collector. @jbw976 touches on this under the "Storage and Durability" section of the design.

jbw976 commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,410 @@
+"configuration drift", it will enforce its "source of truth" to eventually
+correct this unexpected change without any user interaction with the control
+plane.  Currently, there is no mechanism for users of Crossplane to have
+visibility into these types of "unexpected" changes that Crossplane is making on
```
ha i see what you are saying. i'll change the wording slightly to something that states crossplane is reverting the unexpected changes, not making them 😇 

>  Currently, there is no mechanism for users of Crossplane to have visibility into these types of "unexpected" changes that Crossplane is reverting on their behalf.

turkenh reviewed (approved):
Looks good to me, great write-up 🙌

lindblombr commented on design/one-pager-change-logs.md:
```diff
@@ -0,0 +1,432 @@
+| Provider name | `xpkg.upbound.io/upbound/provider-aws-ec2:v1.8.0` |
+| Type (GVK) | `apiVersion: ec2.aws.upbound.io/v1beta2, kind: Instance` |
+| Name | `dev-instance-bt66d` |
+| External Name | `vpc-4fa03d9fb92dfec50` |
```
Any chance we can include resource labels?

Assume we aggregate these changelog events into a search index.  We'd then want to correlate them with higher-level events (ArgoCD or FluxCD resource events) and finally point to a commit (or set of commits).

Having the resource labels as KVs in this schema would allow us to build the relationship with those higher-level events more directly, providing a solid end-to-end view into how changes at the higher-end of the abstraction flow to changes at the lowest end.

vijaymateti commented:
@negz  It would be great if this design doc can be merged as cloud logs already made it to latest release and not much documentation is around for the same. Thank you!
